@claude A few notes:
1. Instead of using bold for key points, use underline
2. Use bold for the headers instead. Strictl don't use markdown headers, because that doesn't get parsed by telegram anyway
3. Also got this error, when trying your changes out.
```
panic: bad request, Bad Request: message is not modified: specified new message content and reply markup are exactly the same as a current content and reply markup of the message
```
4. Also, unrelated to this issue, but for some reason each llm response re-returns the previous message, when it should just send the reply to the new message? I'm not sure if im maintaining the conversation correctly. 
Like if i ask a new question, if the previous replies were irrelevant, it shouldnt re-send those.

--- a/.gitignore
+++ b/.gitignore
@@ -30,3 +30,5 @@ go.work.sum
 # Editor/IDE
 # .idea/
 # .vscode/
+
+conversation.json

--- a/Prompt.md
+++ b/Prompt.md
@@ -1,7 +1,8 @@
 Communicate with extreme brevity and precision. Prioritize accuracy and clarity over verbosity. Provide direct, no-nonsense responses that get straight to the point. When discussing technical topics, especially programming, ensure absolute technical accuracy. Avoid unnecessary elaboration or filler language. Be confident but not obsequious, and focus on delivering concise, high-quality information.
 
 For Telegram messaging:
-- Use **bold** for emphasis on key points
+- Use **bold** for headers and section titles (avoid markdown headers like # ## ###)
+- Use __underline__ for emphasis on key points
 - Use `inline code` for commands, variables, or short code snippets
 - Use code blocks with language specification for longer code examples
 - Structure responses with clear formatting for mobile readability

--- a/Prompt.md
+++ b/Prompt.md
@@ -1,9 +1,27 @@
 Communicate with extreme brevity and precision. Prioritize accuracy and clarity over verbosity. Provide direct, no-nonsense responses that get straight to the point. When discussing technical topics, especially programming, ensure absolute technical accuracy. Avoid unnecessary elaboration or filler language. Be confident but not obsequious, and focus on delivering concise, high-quality information.
 
-For Telegram messaging:
-- Use **bold** for headers and section titles (avoid markdown headers like # ## ###)
-- Use __underline__ for emphasis on key points
-- Use `inline code` for commands, variables, or short code snippets
-- Use code blocks with language specification for longer code examples
-- Structure responses with clear formatting for mobile readability
-- Keep paragraphs short and scannable
+You MUST format all responses using ONLY the HTML tags that Telegram supports. Your response will be parsed as HTML and sent directly to Telegram users.
+<b>bold</b>, <strong>bold</strong>
+<i>italic</i>, <em>italic</em>
+<u>underline</u>, <ins>underline</ins>
+<s>strikethrough</s>, <strike>strikethrough</strike>, <del>strikethrough</del>
+<span class="tg-spoiler">spoiler</span>, <tg-spoiler>spoiler</tg-spoiler>
+<b>bold <i>italic bold <s>italic bold strikethrough <span class="tg-spoiler">italic bold strikethrough spoiler</span></s> <u>underline italic bold</u></i> bold</b>
+<a href="http://www.example.com/">inline URL</a>
+<a href="tg://user?id=123456789">inline mention of a user</a>
+<tg-emoji emoji-id="5368324170671202286">👍</tg-emoji>
+<code>inline fixed-width code</code>
+<pre>pre-formatted fixed-width code block</pre>
+<pre><code class="language-python">pre-formatted fixed-width code block written in the Python programming language</code></pre>
+<blockquote>Block quotation started\nBlock quotation continued\nThe last line of the block quotation</blockquote>
+<blockquote expandable>Expandable block quotation started\nExpandable block quotation continued\nExpandable block quotation continued\nHidden by default part of the block quotation started\nExpandable block quotation continued\nThe last line of the block quotation</blockquote>
+
+Please note:
+
+Only the tags mentioned above are currently supported.
+All <, > and & symbols that are not a part of a tag or an HTML entity must be replaced with the corresponding HTML entities (< with &lt;, > with &gt; and & with &amp;).
+All numerical HTML entities are supported.
+The API currently supports only the following named HTML entities: &lt;, &gt;, &amp; and &quot;.
+Use nested pre and code tags, to define programming language for pre entity.
+Programming language can't be specified for standalone code tags.
+A valid emoji must be used as the content of the tg-emoji tag. The emoji will be shown instead of the custom emoji in places where a custom emoji cannot be displayed (e.g., system notifications) or if the message is forwarded by a non-premium user. It is recommended to use the emoji from the emoji field of the custom emoji sticker.

--- a/main.go
+++ b/main.go
@@ -61,55 +61,141 @@ func loadConversation() ([]anthropic.MessageParam, error) {
 
 // formatForTelegram converts markdown text to Telegram-compatible HTML format
 func formatForTelegram(text string) string {
-	// Escape HTML entities that aren't part of formatting
-	text = strings.ReplaceAll(text, "&", "&amp;")
-	text = strings.ReplaceAll(text, "<", "&lt;")
-	text = strings.ReplaceAll(text, ">", "&gt;")
-
-	// Convert markdown headers to bold text (instead of HTML headers)
-	headerRegex := regexp.MustCompile(`(?m)^#{1,6}\s*(.+)$`)
-	text = headerRegex.ReplaceAllString(text, "<b>$1</b>")
-
-	// Convert code blocks first (to avoid interference with other formatting)
+	// Store code blocks temporarily to avoid processing their content
+	codeBlocks := make(map[string]string)
+	codeBlockCounter := 0
+	
+	// Extract and temporarily replace code blocks
 	codeBlockRegex := regexp.MustCompile("```([a-zA-Z]*)\n?((?s:.*?))\n?```")
 	text = codeBlockRegex.ReplaceAllStringFunc(text, func(match string) string {
 		parts := codeBlockRegex.FindStringSubmatch(match)
 		if len(parts) >= 3 {
 			language := parts[1]
 			code := strings.TrimSpace(parts[2])
+			
+			// Escape HTML in code content
+			code = strings.ReplaceAll(code, "&", "&amp;")
+			code = strings.ReplaceAll(code, "<", "&lt;")
+			code = strings.ReplaceAll(code, ">", "&gt;")
+			
+			placeholder := fmt.Sprintf("__CODEBLOCK_%d__", codeBlockCounter)
 			if language != "" {
-				return fmt.Sprintf("<pre><code class=\"language-%s\">%s</code></pre>", language, code)
+				codeBlocks[placeholder] = fmt.Sprintf("<pre><code class=\"language-%s\">%s</code></pre>", language, code)
+			} else {
+				codeBlocks[placeholder] = fmt.Sprintf("<pre>%s</pre>", code)
 			}
-			return fmt.Sprintf("<pre>%s</pre>", code)
+			codeBlockCounter++
+			return placeholder
 		}
 		return match
 	})
 
-	// Convert inline code
+	// Extract and temporarily replace inline code
+	inlineCodes := make(map[string]string)
+	inlineCodeCounter := 0
 	inlineCodeRegex := regexp.MustCompile("`([^`]+)`")
-	text = inlineCodeRegex.ReplaceAllString(text, "<code>$1</code>")
+	text = inlineCodeRegex.ReplaceAllStringFunc(text, func(match string) string {
+		parts := inlineCodeRegex.FindStringSubmatch(match)
+		if len(parts) >= 2 {
+			code := parts[1]
+			// Escape HTML in inline code
+			code = strings.ReplaceAll(code, "&", "&amp;")
+			code = strings.ReplaceAll(code, "<", "&lt;")
+			code = strings.ReplaceAll(code, ">", "&gt;")
+			
+			placeholder := fmt.Sprintf("__INLINE_%d__", inlineCodeCounter)
+			inlineCodes[placeholder] = fmt.Sprintf("<code>%s</code>", code)
+			inlineCodeCounter++
+			return placeholder
+		}
+		return match
+	})
+
+	// Now escape remaining HTML entities (after extracting code)
+	text = strings.ReplaceAll(text, "&", "&amp;")
+	text = strings.ReplaceAll(text, "<", "&lt;")
+	text = strings.ReplaceAll(text, ">", "&gt;")
+
+	// Convert markdown headers to bold text, but strip existing markdown formatting first
+	headerRegex := regexp.MustCompile(`(?m)^#{1,6}\s*(.+)$`)
+	text = headerRegex.ReplaceAllStringFunc(text, func(match string) string {
+		parts := headerRegex.FindStringSubmatch(match)
+		if len(parts) >= 2 {
+			content := strings.TrimSpace(parts[1])
+			// Strip markdown formatting from headers to avoid double conversion
+			content = regexp.MustCompile(`\*\*([^*]+)\*\*`).ReplaceAllString(content, "$1")
+			content = regexp.MustCompile(`__([^_]+)__`).ReplaceAllString(content, "$1")
+			content = regexp.MustCompile(`\*([^*]+)\*`).ReplaceAllString(content, "$1")
+			content = regexp.MustCompile(`~~([^~]+)~~`).ReplaceAllString(content, "$1")
+			
+			// Don't wrap empty or single character content
+			if len(strings.TrimSpace(content)) > 1 {
+				return fmt.Sprintf("<b>%s</b>", content)
+			}
+			return content
+		}
+		return match
+	})
 
 	// Convert **bold** to bold (for headers and section titles)
-	boldRegex := regexp.MustCompile(`\*\*([^*]+)\*\*`)
+	boldRegex := regexp.MustCompile(`\*\*([^*\n]+)\*\*`)
 	text = boldRegex.ReplaceAllString(text, "<b>$1</b>")
 
 	// Convert __underline__ to underline (for key point emphasis)
-	underlineRegex := regexp.MustCompile(`__([^_]+)__`)
+	underlineRegex := regexp.MustCompile(`__([^_\n]+)__`)
 	text = underlineRegex.ReplaceAllString(text, "<u>$1</u>")
 
-	// Convert italic
-	italicRegex := regexp.MustCompile(`\*([^*]+)\*`)
+	// Convert *italic* but be careful not to match ** patterns
+	italicRegex := regexp.MustCompile(`(?<!\*)\*([^*\n]+)\*(?!\*)`)
 	text = italicRegex.ReplaceAllString(text, "<i>$1</i>")
 
 	// Convert strikethrough (if using ~~text~~)
-	strikethroughRegex := regexp.MustCompile(`~~([^~]+)~~`)
+	strikethroughRegex := regexp.MustCompile(`~~([^~\n]+)~~`)
 	text = strikethroughRegex.ReplaceAllString(text, "<s>$1</s>")
 
 	// Convert links
 	linkRegex := regexp.MustCompile(`\[([^\]]+)\]\(([^)]+)\)`)
 	text = linkRegex.ReplaceAllString(text, "<a href=\"$2\">$1</a>")
 
+	// Restore code blocks
+	for placeholder, code := range codeBlocks {
+		text = strings.ReplaceAll(text, placeholder, code)
+	}
+
+	// Restore inline codes
+	for placeholder, code := range inlineCodes {
+		text = strings.ReplaceAll(text, placeholder, code)
+	}
+
+	// Clean up any malformed HTML or duplicate tags
+	text = cleanupHTML(text)
+
 	return text
+}
+
+// cleanupHTML fixes common HTML issues that could cause Telegram parsing errors
+func cleanupHTML(html string) string {
+	// Remove empty tags
+	emptyTagRegex := regexp.MustCompile(`<(b|i|u|s|code)>\s*</\1>`)
+	html = emptyTagRegex.ReplaceAllString(html, "")
+	
+	// Fix nested identical tags (e.g., <b><b>text</b></b> -> <b>text</b>)
+	for _, tag := range []string{"b", "i", "u", "s", "code"} {
+		nestedRegex := regexp.MustCompile(fmt.Sprintf(`<%s>(\s*<%s>(.+?)</%s>\s*)</%s>`, tag, tag, tag, tag))
+		for nestedRegex.MatchString(html) {
+			html = nestedRegex.ReplaceAllString(html, fmt.Sprintf(`<%s>$2</%s>`, tag, tag))
+		}
+	}
+	
+	// Remove orphaned single characters in tags (like <b>#</b>)
+	orphanRegex := regexp.MustCompile(`<(b|i|u|s)>([^a-zA-Z0-9\s])</\1>`)
+	html = orphanRegex.ReplaceAllString(html, "$2")
+	
+	// Clean up multiple consecutive whitespace while preserving single spaces
+	whitespaceRegex := regexp.MustCompile(`\s+`)
+	html = whitespaceRegex.ReplaceAllString(html, " ")
+	
+	return strings.TrimSpace(html)
 }
 
 func main() {

--- a/main.go
+++ b/main.go
@@ -66,6 +66,10 @@ func formatForTelegram(text string) string {
 	text = strings.ReplaceAll(text, "<", "&lt;")
 	text = strings.ReplaceAll(text, ">", "&gt;")
 
+	// Convert markdown headers to bold text (instead of HTML headers)
+	headerRegex := regexp.MustCompile(`(?m)^#{1,6}\s*(.+)$`)
+	text = headerRegex.ReplaceAllString(text, "<b>$1</b>")
+
 	// Convert code blocks first (to avoid interference with other formatting)
 	codeBlockRegex := regexp.MustCompile("```([a-zA-Z]*)\n?((?s:.*?))\n?```")
 	text = codeBlockRegex.ReplaceAllStringFunc(text, func(match string) string {
@@ -85,17 +89,17 @@ func formatForTelegram(text string) string {
 	inlineCodeRegex := regexp.MustCompile("`([^`]+)`")
 	text = inlineCodeRegex.ReplaceAllString(text, "<code>$1</code>")
 
-	// Convert bold
+	// Convert **bold** to bold (for headers and section titles)
 	boldRegex := regexp.MustCompile(`\*\*([^*]+)\*\*`)
 	text = boldRegex.ReplaceAllString(text, "<b>$1</b>")
+
+	// Convert __underline__ to underline (for key point emphasis)
+	underlineRegex := regexp.MustCompile(`__([^_]+)__`)
+	text = underlineRegex.ReplaceAllString(text, "<u>$1</u>")
 
 	// Convert italic
 	italicRegex := regexp.MustCompile(`\*([^*]+)\*`)
 	text = italicRegex.ReplaceAllString(text, "<i>$1</i>")
-
-	// Convert underline (if using __text__)
-	underlineRegex := regexp.MustCompile(`__([^_]+)__`)
-	text = underlineRegex.ReplaceAllString(text, "<u>$1</u>")
 
 	// Convert strikethrough (if using ~~text~~)
 	strikethroughRegex := regexp.MustCompile(`~~([^~]+)~~`)
@@ -198,6 +202,7 @@ func main() {
 				lastSendTime := time.Now()
 				batchThreshold := 100 * time.Millisecond
 				firstEdit := true
+				lastSentText := ""
 				for stream.Next() {
 					event := stream.Current()
 					err := message.Accumulate(event)
@@ -208,38 +213,44 @@ func main() {
 					now := time.Now()
 					if firstEdit || now.Sub(lastSendTime) >= batchThreshold {
 						if len(message.Content) > 0 && len(message.Content[0].Text) > 0 {
-							if firstEdit {
-								_, err = luffy.SetMessageReaction(ctx, &bot.SetMessageReactionParams{
-									ChatID:    chatID,
-									MessageID: sentMsg.ID,
-									Reaction: []models.ReactionType{
-										{
-											Type: models.ReactionTypeTypeEmoji,
-											ReactionTypeEmoji: &models.ReactionTypeEmoji{
-												Type:  models.ReactionTypeTypeEmoji,
-												Emoji: "✍",
+							formattedText := formatForTelegram(message.Content[0].Text)
+							
+							// Only edit if the content has changed
+							if formattedText != lastSentText {
+								if firstEdit {
+									_, err = luffy.SetMessageReaction(ctx, &bot.SetMessageReactionParams{
+										ChatID:    chatID,
+										MessageID: sentMsg.ID,
+										Reaction: []models.ReactionType{
+											{
+												Type: models.ReactionTypeTypeEmoji,
+												ReactionTypeEmoji: &models.ReactionTypeEmoji{
+													Type:  models.ReactionTypeTypeEmoji,
+													Emoji: "✍",
+												},
 											},
 										},
-									},
+									})
+									if err != nil {
+										log.Panic(err)
+										return
+									}
+									firstEdit = false
+								}
+								
+								_, err = luffy.EditMessageText(ctx, &bot.EditMessageTextParams{
+									ChatID:    chatID,
+									MessageID: sentMsg.ID,
+									Text:      formattedText,
+									ParseMode: models.ParseModeHTML,
 								})
 								if err != nil {
 									log.Panic(err)
 									return
 								}
-								firstEdit = false
+								lastSentText = formattedText
+								lastSendTime = now
 							}
-							formattedText := formatForTelegram(message.Content[0].Text)
-							_, err = luffy.EditMessageText(ctx, &bot.EditMessageTextParams{
-								ChatID:    chatID,
-								MessageID: sentMsg.ID,
-								Text:      formattedText,
-								ParseMode: models.ParseModeHTML,
-							})
-							if err != nil {
-								log.Panic(err)
-								return
-							}
-							lastSendTime = now
 						}
 					}
 				}

--- a/main.go
+++ b/main.go
@@ -2,13 +2,10 @@ package main
 
 import (
 	"context"
-	"encoding/json"
 	"fmt"
 	"log"
 	"os"
 	"os/signal"
-	"regexp"
-	"strings"
 	"syscall"
 	"time"
 
@@ -27,176 +24,7 @@ type BotStats struct {
 	errors         int64
 }
 
-const conversationFile = "conversation.json"
 const promptFile = "Prompt.md"
-
-func saveConversation(conversation []anthropic.MessageParam) error {
-	file, err := os.Create(conversationFile)
-	if err != nil {
-		return err
-	}
-	defer file.Close()
-
-	encoder := json.NewEncoder(file)
-	return encoder.Encode(conversation)
-}
-
-func loadConversation() ([]anthropic.MessageParam, error) {
-	var conversation []anthropic.MessageParam
-
-	file, err := os.Open(conversationFile)
-	if err != nil {
-		if os.IsNotExist(err) {
-			// File doesn't exist, return empty slice
-			return conversation, nil
-		}
-		return nil, err
-	}
-	defer file.Close()
-
-	decoder := json.NewDecoder(file)
-	err = decoder.Decode(&conversation)
-	return conversation, err
-}
-
-// formatForTelegram converts markdown text to Telegram-compatible HTML format
-func formatForTelegram(text string) string {
-	// Store code blocks temporarily to avoid processing their content
-	codeBlocks := make(map[string]string)
-	codeBlockCounter := 0
-	
-	// Extract and temporarily replace code blocks
-	codeBlockRegex := regexp.MustCompile("```([a-zA-Z]*)\n?((?s:.*?))\n?```")
-	text = codeBlockRegex.ReplaceAllStringFunc(text, func(match string) string {
-		parts := codeBlockRegex.FindStringSubmatch(match)
-		if len(parts) >= 3 {
-			language := parts[1]
-			code := strings.TrimSpace(parts[2])
-			
-			// Escape HTML in code content
-			code = strings.ReplaceAll(code, "&", "&amp;")
-			code = strings.ReplaceAll(code, "<", "&lt;")
-			code = strings.ReplaceAll(code, ">", "&gt;")
-			
-			placeholder := fmt.Sprintf("__CODEBLOCK_%d__", codeBlockCounter)
-			if language != "" {
-				codeBlocks[placeholder] = fmt.Sprintf("<pre><code class=\"language-%s\">%s</code></pre>", language, code)
-			} else {
-				codeBlocks[placeholder] = fmt.Sprintf("<pre>%s</pre>", code)
-			}
-			codeBlockCounter++
-			return placeholder
-		}
-		return match
-	})
-
-	// Extract and temporarily replace inline code
-	inlineCodes := make(map[string]string)
-	inlineCodeCounter := 0
-	inlineCodeRegex := regexp.MustCompile("`([^`]+)`")
-	text = inlineCodeRegex.ReplaceAllStringFunc(text, func(match string) string {
-		parts := inlineCodeRegex.FindStringSubmatch(match)
-		if len(parts) >= 2 {
-			code := parts[1]
-			// Escape HTML in inline code
-			code = strings.ReplaceAll(code, "&", "&amp;")
-			code = strings.ReplaceAll(code, "<", "&lt;")
-			code = strings.ReplaceAll(code, ">", "&gt;")
-			
-			placeholder := fmt.Sprintf("__INLINE_%d__", inlineCodeCounter)
-			inlineCodes[placeholder] = fmt.Sprintf("<code>%s</code>", code)
-			inlineCodeCounter++
-			return placeholder
-		}
-		return match
-	})
-
-	// Now escape remaining HTML entities (after extracting code)
-	text = strings.ReplaceAll(text, "&", "&amp;")
-	text = strings.ReplaceAll(text, "<", "&lt;")
-	text = strings.ReplaceAll(text, ">", "&gt;")
-
-	// Convert markdown headers to bold text, but strip existing markdown formatting first
-	headerRegex := regexp.MustCompile(`(?m)^#{1,6}\s*(.+)$`)
-	text = headerRegex.ReplaceAllStringFunc(text, func(match string) string {
-		parts := headerRegex.FindStringSubmatch(match)
-		if len(parts) >= 2 {
-			content := strings.TrimSpace(parts[1])
-			// Strip markdown formatting from headers to avoid double conversion
-			content = regexp.MustCompile(`\*\*([^*]+)\*\*`).ReplaceAllString(content, "$1")
-			content = regexp.MustCompile(`__([^_]+)__`).ReplaceAllString(content, "$1")
-			content = regexp.MustCompile(`\*([^*]+)\*`).ReplaceAllString(content, "$1")
-			content = regexp.MustCompile(`~~([^~]+)~~`).ReplaceAllString(content, "$1")
-			
-			// Don't wrap empty or single character content
-			if len(strings.TrimSpace(content)) > 1 {
-				return fmt.Sprintf("<b>%s</b>", content)
-			}
-			return content
-		}
-		return match
-	})
-
-	// Convert **bold** to bold (for headers and section titles)
-	boldRegex := regexp.MustCompile(`\*\*([^*\n]+)\*\*`)
-	text = boldRegex.ReplaceAllString(text, "<b>$1</b>")
-
-	// Convert __underline__ to underline (for key point emphasis)
-	underlineRegex := regexp.MustCompile(`__([^_\n]+)__`)
-	text = underlineRegex.ReplaceAllString(text, "<u>$1</u>")
-
-	// Convert *italic* but be careful not to match ** patterns
-	italicRegex := regexp.MustCompile(`(?<!\*)\*([^*\n]+)\*(?!\*)`)
-	text = italicRegex.ReplaceAllString(text, "<i>$1</i>")
-
-	// Convert strikethrough (if using ~~text~~)
-	strikethroughRegex := regexp.MustCompile(`~~([^~\n]+)~~`)
-	text = strikethroughRegex.ReplaceAllString(text, "<s>$1</s>")
-
-	// Convert links
-	linkRegex := regexp.MustCompile(`\[([^\]]+)\]\(([^)]+)\)`)
-	text = linkRegex.ReplaceAllString(text, "<a href=\"$2\">$1</a>")
-
-	// Restore code blocks
-	for placeholder, code := range codeBlocks {
-		text = strings.ReplaceAll(text, placeholder, code)
-	}
-
-	// Restore inline codes
-	for placeholder, code := range inlineCodes {
-		text = strings.ReplaceAll(text, placeholder, code)
-	}
-
-	// Clean up any malformed HTML or duplicate tags
-	text = cleanupHTML(text)
-
-	return text
-}
-
-// cleanupHTML fixes common HTML issues that could cause Telegram parsing errors
-func cleanupHTML(html string) string {
-	// Remove empty tags
-	emptyTagRegex := regexp.MustCompile(`<(b|i|u|s|code)>\s*</\1>`)
-	html = emptyTagRegex.ReplaceAllString(html, "")
-	
-	// Fix nested identical tags (e.g., <b><b>text</b></b> -> <b>text</b>)
-	for _, tag := range []string{"b", "i", "u", "s", "code"} {
-		nestedRegex := regexp.MustCompile(fmt.Sprintf(`<%s>(\s*<%s>(.+?)</%s>\s*)</%s>`, tag, tag, tag, tag))
-		for nestedRegex.MatchString(html) {
-			html = nestedRegex.ReplaceAllString(html, fmt.Sprintf(`<%s>$2</%s>`, tag, tag))
-		}
-	}
-	
-	// Remove orphaned single characters in tags (like <b>#</b>)
-	orphanRegex := regexp.MustCompile(`<(b|i|u|s)>([^a-zA-Z0-9\s])</\1>`)
-	html = orphanRegex.ReplaceAllString(html, "$2")
-	
-	// Clean up multiple consecutive whitespace while preserving single spaces
-	whitespaceRegex := regexp.MustCompile(`\s+`)
-	html = whitespaceRegex.ReplaceAllString(html, " ")
-	
-	return strings.TrimSpace(html)
-}
 
 func main() {
 	tgBotToken := os.Getenv("TG_BOT_TOKEN")
@@ -216,8 +44,14 @@ func main() {
 
 	opts := []bot.Option{
 		bot.WithDefaultHandler(func(ctx context.Context, b *bot.Bot, update *models.Update) {
-			tgStream <- update
-			// TODO: close the channel, when we get a context update
+			select {
+			case tgStream <- update:
+				// Successfully sent update
+			case <-ctx.Done():
+				// Context cancelled, close channel and return
+				close(tgStream)
+				return
+			}
 		}),
 	}
 
@@ -230,21 +64,7 @@ func main() {
 	go luffy.Start(ctx)
 
 	ai := anthropic.NewClient()
-	conversation, err := loadConversation()
-	if err != nil {
-		log.Printf("Error loading conversation: %v", err)
-		conversation = []anthropic.MessageParam{}
-	} else {
-		fmt.Printf("Loaded %d messages from previous session\n", len(conversation))
-	}
-
-	defer func() {
-		if err := saveConversation(conversation); err != nil {
-			log.Printf("Error saving conversation: %v", err)
-		} else {
-			fmt.Printf("Saved %d messages\n", len(conversation))
-		}
-	}()
+	conversation := []anthropic.MessageParam{}
 
 	promptTxt, err := os.ReadFile(promptFile)
 	if err != nil {
@@ -286,7 +106,7 @@ func main() {
 
 				message := anthropic.Message{}
 				lastSendTime := time.Now()
-				batchThreshold := 100 * time.Millisecond
+				batchThreshold := 1 * time.Second
 				firstEdit := true
 				lastSentText := ""
 				for stream.Next() {
@@ -299,10 +119,10 @@ func main() {
 					now := time.Now()
 					if firstEdit || now.Sub(lastSendTime) >= batchThreshold {
 						if len(message.Content) > 0 && len(message.Content[0].Text) > 0 {
-							formattedText := formatForTelegram(message.Content[0].Text)
-							
+							text := message.Content[0].Text
+
 							// Only edit if the content has changed
-							if formattedText != lastSentText {
+							if text != lastSentText {
 								if firstEdit {
 									_, err = luffy.SetMessageReaction(ctx, &bot.SetMessageReactionParams{
 										ChatID:    chatID,
@@ -323,28 +143,48 @@ func main() {
 									}
 									firstEdit = false
 								}
-								
+
 								_, err = luffy.EditMessageText(ctx, &bot.EditMessageTextParams{
 									ChatID:    chatID,
 									MessageID: sentMsg.ID,
-									Text:      formattedText,
-									ParseMode: models.ParseModeHTML,
+									Text:      text,
 								})
 								if err != nil {
 									log.Panic(err)
 									return
 								}
-								lastSentText = formattedText
+								lastSentText = text
 								lastSendTime = now
 							}
 						}
 					}
 				}
-				conversation = append(conversation, message.ToParam())
+
+				// Send final message with HTML parsing if content is complete
+				finalText := ""
+				if len(message.Content) > 0 {
+					finalText = message.Content[0].Text
+				}
+
+				_, err = luffy.EditMessageText(ctx, &bot.EditMessageTextParams{
+					ChatID:    chatID,
+					MessageID: sentMsg.ID,
+					Text:      finalText,
+					ParseMode: models.ParseModeHTML,
+				})
+				if err != nil {
+					log.Panic(err)
+					return
+				}
 
 				if stream.Err() != nil {
 					log.Panic(stream.Err())
 					continue
+				}
+
+				// Only append message if it has content
+				if len(message.Content) > 0 {
+					conversation = append(conversation, message.ToParam())
 				}
 
 				_, err = luffy.SetMessageReaction(ctx, &bot.SetMessageReactionParams{
@@ -362,58 +202,4 @@ func main() {
 			return
 		}
 	}
-
-	// for update := range updates {
-	// 	if update.Message != nil {
-	// 		log.Printf("[%s] %s", update.Message.From.UserName, update.Message.Text)
-
-	// 		chatID := update.Message.Chat.ID
-
-	// 		tgMsg := tgbotapi.NewMessage(chatID, "")
-	// 		tgMsg.ReplyToMessageID = update.Message.MessageID
-	// 		sentMsg, err := bot.Send(tgMsg)
-	// 		if err != nil {
-	// 			log.Panic(err)
-	// 			break
-	//
-	//
-	// 		// // Make the call to the anthropic api servers
-	// 		// message, err := ai.Messages.New(context.TODO(), anthropic.MessageNewParams{
-	// 		// 	Model:     anthropic.ModelClaude4Sonnet20250514,
-	// 		// 	MaxTokens: int64(1024),
-	// 		// 	Messages:  conversation,
-	// 		// 	System: systemPrompt,
-	// 		// })
-	// 		// if err != nil {
-	// 		// 	log.Panic(err)
-	// 		// }
-	// 		// conversation = append(conversation, message.ToParam())
-
-	// 		// for _, content := range message.Content {
-	// 		// 	switch content.Type {
-	// 		// 	case "text":
-	// 		// 		msg := tgbotapi.NewMessage(update.Message.Chat.ID, content.Text)
-	// 		// 		msg.ReplyToMessageID = update.Message.MessageID
-
-	// 		// 		// Delete the thinking Message first
-	// 		// 		deleteMsg := tgbotapi.NewDeleteMessage(msg.ChatID, sentMsg.MessageID)
-	// 		// 		bot.Send(deleteMsg)
-
-	// 		// 		bot.Send(msg)
-	// 		// 	}
-	// 		// }
-	// 	}
-	// }
-}
-
-func NewAgent(client *anthropic.Client, getUserMessage func() (string, bool)) *Agent {
-	return &Agent{
-		client:         client,
-		getUserMessage: getUserMessage,
-	}
-}
-
-type Agent struct {
-	client         *anthropic.Client
-	getUserMessage func() (string, bool)
 }

